docs/tutorial/creating_cli.md

```diff
@@ -0,0 +1,622 @@
+          )
+          
+end addContact
+```
+
+First, we read all of our lines, line by line using the `FilesOs#readLines` method (1). Then, we check if the contact already exists in the file checking if the username is contained (2). If it does, we raise an error using the `raiseError` method. If it doesn't, we check if the file is empty (3) using the `ifM`; this method basically checks if the boolean inside the `IO` is true or false and it executes an effect depending on the condition. If it is, we write the contact to the file using the `write` method. If it's not, we append the contact to the file using the `appendLine` method (4).
+
+
+Now that we have the `addContact` function implemented, we can move to the `removeContact` function:
+
+```scala 3
+override def removeContact(username: Username): IO[Unit] =
+  for
+    contacts <- bookPath.readLines
+    _ <- bookPath.writeLines(
+      contacts.filterNot(_.startsWith(username))
+    )
+  yield ()
+```
+
+The explanation is straightforward. We load all the contacts from the file, and then we filter the ones that don't match the username we want to remove. Finally, we write the filtered contacts back to the file.
+
+Now is the turn of the search functions. The first one is the `searchId` function:
+
+```scala 3
+override def searchId(username: Username): IO[Option[Contact]] =
+  for 
+    contacts <- bookPath.readLines
+  yield contacts
+    .find(_.startsWith(username))
+    .flatMap(parseContact(_).toOption)
+```
+
+Similar to read, we load all the contacts from the file, and then we find the contact that matches the username we want to search using the `find` functions, it returns an `Option` whenever it found something or not. If we find it, we parse it to a `Contact` and bind it as an `Option`.
+
+The next three functions are very similar, so we will only show one of them, the `searchName` function:
+
+```scala 3
+override def searchName(name: Name): IO[List[Contact]] =
+  for
+    contacts <- bookPath.readLines.map(_.filter(_.contains(name)))
+    parsed <- contacts.traverse(contact => 
+      IO.fromEither(parseContact(contact)) 
+    )
+  yield parsed
+```
+
+The first part is similar to the previous functions, we load all the contacts from the file, and then we filter the ones that contain the name (or field) we want to search. Look! It's our friend `traverse` again! We're going to use it to parse the file representation of the contacts to our `Contact`using the `fromEither` method, as this will raise the error if it encounters a `Left` inside the `IO`.
+
+Next, we have the `getAll` function to list all the contacts:
+
+```scala 3
+override def getAll: IO[List[Contact]] =
+  for
+    contacts <- bookPath.readLines
+    parsed <- contacts.traverse(contact =>
+      IO.fromEither(parseContact(contact))
+    )
+  yield parsed
+```
+In like the search functions, we load all the contacts from the file, and then we parse them to a `Contact` using the `traverse` method.
```

Comment:
```suggestion
Just like the search functions, we load all the contacts from the file, and then we parse them to a `Contact` using the `traverse` method.
```
The wording here is also a bit iffy, `traverse` doesn't parse it, `parseContact` does it, but `traverse` makes it so that we return a `IO[List[Contact]]` and not a `List[IO[Contact]]`. I think it's important to convey that `traverse` actually does that.

Reply:
You are right!